File: src/nbmetis.cpp
// METIS bindings for Python using Nanobind

#include <metis.h>
#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <nanobind/stl/pair.h>

namespace nb = nanobind;
using namespace nb::literals;

using IdxA = nb::ndarray<idx_t, nb::shape<-1>, nb::device::cpu>;
using RealA = nb::ndarray<real_t, nb::shape<-1>, nb::device::cpu>;

// (Return value, edgecut)
using PartGraphRet = std::pair<int, idx_t>;

PartGraphRet PartGraph(idx_t nvtxs, idx_t ncon, idx_t nparts, IdxA xadj,
                       IdxA adjncy, IdxA vwgt, IdxA vsize, IdxA adjwgt,
                       RealA tpwgts, RealA ubvec, IdxA options, IdxA part,
                       int part_kway) {
  idx_t edgecut = 0;

  idx_t *nvtxs_p = &nvtxs;
  idx_t *ncon_p = &ncon;

  idx_t *xadj_p = xadj.data();
  idx_t *adjncy_p = adjncy.data();

  idx_t *vwgt_p = vwgt.size() == 0 ? nullptr : vwgt.data();
  idx_t *vsize_p = vsize.size() == 0 ? nullptr : vsize.data();
  idx_t *adjwgt_p = adjwgt.size() == 0 ? nullptr : adjwgt.data();

  idx_t *nparts_p = &nparts;

  real_t *tpwgts_p = tpwgts.size() == 0 ? nullptr : tpwgts.data();
  real_t *ubvec_p = ubvec.size() == 0 ? nullptr : ubvec.data();
  idx_t *options_p = options.size() == 0 ? nullptr : options.data();

  idx_t *edgecut_p = &edgecut;
  idx_t *part_p = part.data();

  int ret;
  if (part_kway) {
    ret = METIS_PartGraphKway(nvtxs_p, ncon_p, xadj_p, adjncy_p, vwgt_p,
                              vsize_p, adjwgt_p, nparts_p, tpwgts_p, ubvec_p,
                              options_p, edgecut_p, part_p);
  } else {
    ret = METIS_PartGraphRecursive(nvtxs_p, ncon_p, xadj_p, adjncy_p, vwgt_p,
                                   vsize_p, adjwgt_p, nparts_p, tpwgts_p,
                                   ubvec_p, options_p, edgecut_p, part_p);
  }

  return PartGraphRet(ret, edgecut);
}

int NodeND(idx_t nvtxs, IdxA xadj, IdxA adjncy, IdxA vwgt, IdxA options,
           IdxA perm, IdxA iperm) {

  // std::cout << "nvtxs = " << nvtxs << "\n";
  // std::cout << "xadj.size() = " << xadj.size() << "\n";
  // std::cout << "adjncy.size() = " << adjncy.size() << "\n";
  // std::cout << "vwgt.size() = " << vwgt.size() << "\n";
  // std::cout << "options.size() = " << options.size() << "\n";
  // std::cout << "perm.size() = " << perm.size() << "\n";
  // std::cout << "iperm.size() = " << iperm.size() << "\n";

  idx_t *nvtxs_p = &nvtxs;
  idx_t *xadj_p = xadj.data();
  idx_t *adjncy_p = adjncy.data();
  idx_t *vwgt_p = vwgt.size() == 0 ? nullptr : vwgt.data();
  idx_t *options_p = options.size() == 0 ? nullptr : options.data();

  idx_t *perm_p = perm.data();
  idx_t *iperm_p = iperm.data();

  // for (int i = 0; i < nvtxs; i++) {
  //     int start = xadj_p[i];
  //     int end = xadj_p[i+1];
  //     for (int j = start; j < end; j++) {
  //         std::cout << "edge" << i << " " << adjncy_p[j] << "\n";
  //     }
  // }

  return METIS_NodeND(nvtxs_p, xadj_p, adjncy_p, vwgt_p, options_p, perm_p,
                      iperm_p);
}

int SetDefaultOptions(IdxA options) {
  return METIS_SetDefaultOptions(options.data());
}

#define EXPORT_CONSTANT(m, a) m.attr(#a) = int(METIS_##a)

NB_MODULE(_nbmetis, m) {
  m.doc() = "Yet anothor Python binding for METIS";

  m.attr("SIZEOF_IDX_T") = sizeof(idx_t);
  m.attr("SIZEOF_REAL_T") = sizeof(real_t);

  EXPORT_CONSTANT(m, VER_MAJOR);
  EXPORT_CONSTANT(m, VER_MINOR);
  EXPORT_CONSTANT(m, VER_SUBMINOR);

  EXPORT_CONSTANT(m, NOPTIONS);

  EXPORT_CONSTANT(m, OK);
  EXPORT_CONSTANT(m, ERROR_INPUT);
  EXPORT_CONSTANT(m, ERROR_MEMORY);
  EXPORT_CONSTANT(m, ERROR);

  EXPORT_CONSTANT(m, OP_PMETIS);
  EXPORT_CONSTANT(m, OP_KMETIS);
  EXPORT_CONSTANT(m, OP_OMETIS);

  EXPORT_CONSTANT(m, OPTION_PTYPE);
  EXPORT_CONSTANT(m, OPTION_OBJTYPE);
  EXPORT_CONSTANT(m, OPTION_CTYPE);
  EXPORT_CONSTANT(m, OPTION_IPTYPE);
  EXPORT_CONSTANT(m, OPTION_RTYPE);
  EXPORT_CONSTANT(m, OPTION_DBGLVL);
  EXPORT_CONSTANT(m, OPTION_NITER);
  EXPORT_CONSTANT(m, OPTION_NCUTS);
  EXPORT_CONSTANT(m, OPTION_SEED);
  EXPORT_CONSTANT(m, OPTION_NO2HOP);
  EXPORT_CONSTANT(m, OPTION_MINCONN);
  EXPORT_CONSTANT(m, OPTION_CONTIG);
  EXPORT_CONSTANT(m, OPTION_COMPRESS);
  EXPORT_CONSTANT(m, OPTION_CCORDER);
  EXPORT_CONSTANT(m, OPTION_PFACTOR);
  EXPORT_CONSTANT(m, OPTION_NSEPS);
  EXPORT_CONSTANT(m, OPTION_UFACTOR);
  EXPORT_CONSTANT(m, OPTION_NUMBERING);

  EXPORT_CONSTANT(m, PTYPE_RB);
  EXPORT_CONSTANT(m, PTYPE_KWAY);

  EXPORT_CONSTANT(m, GTYPE_DUAL);
  EXPORT_CONSTANT(m, GTYPE_NODAL);

  EXPORT_CONSTANT(m, CTYPE_RM);
  EXPORT_CONSTANT(m, CTYPE_SHEM);

  EXPORT_CONSTANT(m, IPTYPE_GROW);
  EXPORT_CONSTANT(m, IPTYPE_RANDOM);
  EXPORT_CONSTANT(m, IPTYPE_EDGE);
  EXPORT_CONSTANT(m, IPTYPE_NODE);
  EXPORT_CONSTANT(m, IPTYPE_METISRB);

  EXPORT_CONSTANT(m, RTYPE_FM);
  EXPORT_CONSTANT(m, RTYPE_GREEDY);
  EXPORT_CONSTANT(m, RTYPE_SEP2SIDED);
  EXPORT_CONSTANT(m, RTYPE_SEP1SIDED);

  EXPORT_CONSTANT(m, DBG_INFO);
  EXPORT_CONSTANT(m, DBG_TIME);
  EXPORT_CONSTANT(m, DBG_COARSEN);
  EXPORT_CONSTANT(m, DBG_REFINE);
  EXPORT_CONSTANT(m, DBG_IPART);
  EXPORT_CONSTANT(m, DBG_MOVEINFO);
  EXPORT_CONSTANT(m, DBG_SEPINFO);
  EXPORT_CONSTANT(m, DBG_CONNINFO);
  EXPORT_CONSTANT(m, DBG_CONTIGINFO);
  EXPORT_CONSTANT(m, DBG_MEMORY);

  EXPORT_CONSTANT(m, OBJTYPE_CUT);
  EXPORT_CONSTANT(m, OBJTYPE_VOL);
  EXPORT_CONSTANT(m, OBJTYPE_NODE);

  m.def("PartGraph", &PartGraph, "nvtxs"_a, "ncon"_a, "nparts"_a, "xadj"_a,
        "adjncy"_a, "vwgt"_a, "vsize"_a, "adjwgt"_a, "tpwgts"_a, "ubvec"_a,
        "options"_a, "part"_a, "part_kway"_a,
        "Is used to partition a graph into k parts using either multilevel "
        "recursive bisection or multilevel k-way partitioning.");

  m.def("NodeND", &NodeND, "nvtxs"_a, "xadj"_a, "adjncy"_a, "vwgt"_a,
        "options"_a, "perm"_a, "iperm"_a,
        "This function computes fill reducing orderings of sparse matrices "
        "using the multilevel nested dissection algorithm.");

  m.def("SetDefaultOptions", &SetDefaultOptions, "options"_a,
        "Initializes the options array into its default values.");
}

File: src/nbmetis/nbmetis.py
"""Low level bindings."""

import random
from typing import Annotated

from . import _nbmetis as metis

import numpy as np
from numpy import ndarray
from scipy.sparse import csr_matrix

if metis.SIZEOF_IDX_T == 64:
    idx_t = np.int64
    idx_t_dtype_str = "int64"
else:
    idx_t = np.int32
    idx_t_dtype_str = "int32"

if metis.SIZEOF_REAL_T == 64:
    real_t = np.float32
    real_t_dtype_str = "float32"
else:
    real_t = np.float64
    real_t_dtype_str = "float64"

IdxA = Annotated[ndarray, dict(dtype=idx_t_dtype_str, shape=(None), device="cpu")]
RealA = Annotated[ndarray, dict(dtype=real_t_dtype_str, shape=(None), device="cpu")]

IdxA_NULL: IdxA = np.zeros(shape=(0,), dtype=idx_t)
RealA_NULL: RealA = np.zeros(shape=(0,), dtype=real_t)


def part_graph(
    num_nodes: int,
    num_edges: int,
    num_parts: int,
    weighted_adjacency: csr_matrix,
    vertex_weights: ndarray | None,
    vertex_size: ndarray | None,
    target_parition_weights: ndarray | None,
    allowed_load_imbalance: ndarray | None,
    options: ndarray | None,
    part_kway: bool,
) -> tuple[int, ndarray]:
    assert weighted_adjacency.shape == (num_nodes, num_nodes)
    assert weighted_adjacency.indptr.shape == (num_nodes + 1,)
    assert weighted_adjacency.indices.shape == (2 * num_edges,)

    if vertex_weights is not None:
        assert len(vertex_weights.shape) == 2
        assert vertex_weights.shape[0] == num_nodes
    if vertex_size is not None:
        assert vertex_size.shape == (num_nodes,)

    xadj = weighted_adjacency.indptr.astype(idx_t, copy=False)
    adjncy = weighted_adjacency.indices.astype(idx_t, copy=False)
    if vertex_weights is None:
        vwgt = IdxA_NULL
    else:
        vwgt = vertex_weights.astype(idx_t, copy=False)
    if vertex_size is None:
        vsize = IdxA_NULL
    else:
        vsize = vertex_size.astype(idx_t, copy=False)

    adjwgt = weighted_adjacency.data.astype(idx_t, copy=False)

    if target_parition_weights is None:
        tpwgts = RealA_NULL
    else:
        tpwgts = target_parition_weights.astype(real_t, copy=False)
    if allowed_load_imbalance is None:
        ubvec = RealA_NULL
    else:
        ubvec = allowed_load_imbalance.astype(real_t, copy=False)

    if options is None:
        options = IdxA_NULL
    else:
        options = options.astype(idx_t, copy=False)

    part = np.zeros(shape=(num_nodes,), dtype=idx_t)

    ret, edgecut = metis.PartGraph(
        nparts=num_parts,
        xadj=xadj,
        adjncy=adjncy,
        vwgt=vwgt,
        vsize=vsize,
        adjwgt=adjwgt,
        tpwgts=tpwgts,
        ubvec=ubvec,
        options=options,
        part=part,
        part_kway=int(part_kway),
    )

    assert ret == metis.OK

    return edgecut, part


def node_nd(
    num_nodes: int,
    num_edges: int,
    weighted_adjacency: csr_matrix,
    vertex_weights: ndarray | None,
    options: ndarray | None,
) -> tuple[ndarray, ndarray]:
    assert weighted_adjacency.shape == (num_nodes, num_nodes)
    assert weighted_adjacency.indptr.shape == (num_nodes + 1,)
    assert weighted_adjacency.indices.shape == (2 * num_edges,)

    if vertex_weights is not None:
        assert vertex_weights.shape == (num_nodes,)

    xadj = weighted_adjacency.indptr.astype(idx_t, copy=False)
    adjncy = weighted_adjacency.indices.astype(idx_t, copy=False)

    if vertex_weights is None:
        vwgt = IdxA_NULL
    else:
        vwgt = vertex_weights.astype(idx_t, copy=False)

    if options is None:
        options = IdxA_NULL
    else:
        options = options.astype(idx_t, copy=False)

    perm = np.zeros(shape=(num_nodes,), dtype=idx_t)
    iperm = np.zeros(shape=(num_nodes,), dtype=idx_t)

    ret = metis.NodeND(
        nvtxs=num_nodes,
        xadj=xadj,
        adjncy=adjncy,
        vwgt=vwgt,
        options=options,
        perm=perm,
        iperm=iperm,
    )
    assert ret == metis.OK

    return perm, iperm


def default_options_node_nd(verbose: bool = False) -> ndarray:
    options = np.zeros(shape=(metis.NOPTIONS), dtype=idx_t)
    ret = metis.SetDefaultOptions(options)
    assert ret == metis.OK

    options[metis.OPTION_OBJTYPE] = metis.OBJTYPE_NODE
    options[metis.OPTION_PTYPE] = metis.PTYPE_KWAY
    options[metis.OPTION_CTYPE] = metis.CTYPE_SHEM
    options[metis.OPTION_IPTYPE] = metis.IPTYPE_EDGE
    options[metis.OPTION_RTYPE] = metis.RTYPE_SEP1SIDED
    options[metis.OPTION_UFACTOR] = 30
    options[metis.OPTION_PFACTOR] = 0
    options[metis.OPTION_NO2HOP] = 0
    options[metis.OPTION_COMPRESS] = 1
    options[metis.OPTION_CCORDER] = 1
    options[metis.OPTION_NITER] = 10
    options[metis.OPTION_NSEPS] = 1
    options[metis.OPTION_NUMBERING] = 0
    options[metis.OPTION_SEED] = random.randint(0, 2**31 - 1)
    if verbose:
        options[metis.OPTION_DBGLVL] = metis.DBG_INFO
    else:
        options[metis.OPTION_DBGLVL] = 0


    return options
